### PR TITLE
Refactor Optimiser Steps Interface.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libyul/optimiser/OptimiserStep.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+	virtual void run(OptimiserStepContext&, Block&) const = 0;
+	std::string name;
+};
+
+template <class Step>
+struct OptimiserStepInstance: public OptimiserStep
+{
+	OptimiserStepInstance(): OptimiserStep{Step::name} {}
+	virtual void run(OptimiserStepContext& _context, Block& _ast) const override
```
Comment (leonardoalt) [inline libyul/optimiser/OptimiserStep.h]: Didn't we agree at some point that these would be `override` only?

**Comment by leonardoalt** [comment]
Looks good, but build failing

**Comment by bshastry** [comment]
Via clang v10.0

> SSAReverser.h:80:2: error: explicitly defaulted default constructor is implicitly deleted [-Werror,-Wdefaulted-function-deleted]

Looks like it's time for us to use a standalone clang CI :-)

File: libyul/optimiser/OptimiserStep.h (hunk 1 above)
Comment (chriseth) [reply libyul/optimiser/OptimiserStep.h]: Yep!

**Comment by chriseth** [comment]
Updated.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**